### DISCUSSION TITLE
Using docker-mailserver with IPv6

### DISCUSSION THREAD

**Discussion by some-user123** [docker-mailserver/docker-mailserver#2912] [Q&A]
I'm using the experimental IPv6 support from docker on my host machine and avoid using  `robbertkl/ipv6nat`. Userland-proxy is disabled, ip6tables in enabled. This works great for all my containers, except docker-mailserver :-/

I'm having an IPv6 enabled network and the relevant ports are exposed on both IPv4 and IPv6:
```
860bbf1f042e   mailserver/docker-mailserver:11.2                 "/usr/bin/dumb-init …"   53 minutes ago   Up 53 minutes           0.0.0.0:25->25/tcp, :::25->25/tcp, 110/tcp, 0.0.0.0:143->143/tcp, :::143->143/tcp, ...   mailserver
```

However, I cannot connect:
```
root@ex:~# nc -6 example.com 143 -v
nc: connect to example.com port 143 (tcp) failed: Connection refused
root@ex:~# nc -4 example.com 143 -v
Connection to example.com 143 port [tcp/imap2] succeeded!
* OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE LITERAL+ STARTTLS LOGINDISABLED] Dovecot (Debian) ready.
```

I think, the reason is that the dovecot daemon is configured in `dovecot.conf` to listen on IPv4 only:
```
listen = *
```

(Haven't checked postfix yet, but I guess it's similar there...)

Is there a way to let dovecot and postfix listen on IPv6 addresses as well that are properly configured for the container (`listen = *, ::` should do the trick for dovecot) without messing and replacing the complete dovecot/postfix config?

I also don't get how this should be working with `robbertkl/ipv6nat` following the instructions here: https://docker-mailserver.github.io/docker-mailserver/v11.2/config/advanced/ipv6/. Postfix/Dovecot won't still not listen on IPv6, or why would they?

**Comment by MichiMeyer** [comment]
> I also don't get how this should be working with robbertkl/ipv6nat following the instructions here: https://docker-mailserver.github.io/docker-mailserver/v11.2/config/advanced/ipv6/. Postfix/Dovecot won't still not listen on IPv6, or why would they?

Have it running with ipv6nat docker and some privat ipv6-subnet. I don't like using it that way, but I had limited time while setting it up. I would be *really* interested in having all the stack running with *native* ipv6 and without any nat involved.

I think the basic ipv6nat design was like "using the hosts ipv6 and using a private ipv6 for dockers" then doing bridging and natting for getting the packets from a to b. The Service inside the docker are listening to both ipv4+ipv6. The docker-proxy just ipv4. The mailserver.env brought the ipv6-part inside the docker up if I remember correctly.

docker-compose.(override):
```
  ipv6nat:
    image: robbertkl/ipv6nat
    restart: always
    network_mode: "host"
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /lib/modules:/lib/modules:ro
#[..]
networks:
  default:
    driver: bridge
    enable_ipv6: true
    ipam:
      driver: default
      config:
        - subnet: fd00:0123:4567::/48
          gateway: fd00:0123:4567::1
```

Which gives some private ipv6 to the dockers:
```# ip -6 neigh | grep fd
fd00:123:4567::2 dev br-558b04a9a278 lladdr 02:42:ac:12:00:02 STALE
fd00:123:4567::6 dev br-558b04a9a278 lladdr 02:42:ac:12:00:06 DELAY
fd00:123:4567::4 dev br-558b04a9a278 lladdr 02:42:ac:12:00:04 STALE
# ip -6 a s dev br-558b04a9a278
36: br-558b04a9a278: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    inet6 fd00:123:4567::1/48 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::42:45ff:fe49:13bc/64 scope link 
       valid_lft forever preferred_lft forever
    inet6 fe80::1/64 scope link 
       valid_lft forever preferred_lft forever
```

Looking at the ip6tables, there are some rules for accept and NAT as DNAT and SNAT.
```-A DOCKER -d fd00:123:4567::2/128 ! -i br-558b04a9a278 -o br-558b04a9a278 -p tcp -m tcp --dport 25 -j ACCEPT
-A POSTROUTING -s fd00:123:4567::2/128 -d fd00:123:4567::2/128 -p tcp -m tcp --dport 25 -j MASQUERADE
-A DOCKER ! -i br-558b04a9a278 -p tcp -m tcp --dport 25 -j DNAT --to-destination [fd00:123:4567::2]:25
```

While the docker-proxy is listening on ipv4 only:
```# netstat -tulpen | grep :25
tcp        0      0 0.0.0.0:25              0.0.0.0:*               LISTEN      0          328169     111139/docker-proxy
```
The docker itself uses v6:
```# telnet fd00:123:4567::2 25
Trying fd00:123:4567::2...
Connected to fd00:123:4567::2.
Escape character is '^]'.
220-xxxx ESMTP
```

mailserver.env has `POSTFIX_INET_PROTOCOLS` and `DOVECOT_INET_PROTOCOLS=all` set.

Inside the docker the services are listening for ipv4 and ipv6 (shortend output):
```
root@mail01:/# netstat -tulpen
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       User       Inode      PID/Program name    
tcp        0      0 0.0.0.0:143             0.0.0.0:*               LISTEN      0          383772     15787/dovecot       
tcp        0      0 0.0.0.0:25              0.0.0.0:*               LISTEN      0          384890     -                   
tcp        0      0 0.0.0.0:465             0.0.0.0:*               LISTEN      0          384911     -                   
tcp        0      0 0.0.0.0:587             0.0.0.0:*               LISTEN      0          384905     -                   
tcp        0      0 0.0.0.0:993             0.0.0.0:*               LISTEN      0          383774     15787/dovecot       
tcp        0      0 0.0.0.0:9900            0.0.0.0:*               LISTEN      0          383734     15787/dovecot       
tcp6       0      0 :::143                  :::*                    LISTEN      0          383773     15787/dovecot       
tcp6       0      0 :::25                   :::*                    LISTEN      0          384891     -                   
tcp6       0      0 :::4190                 :::*                    LISTEN      0          383726     15787/dovecot       
tcp6       0      0 :::465                  :::*                    LISTEN      0          384912     -                   
tcp6       0      0 :::587                  :::*                    LISTEN      0          384906     -                   
tcp6       0      0 :::993                  :::*                    LISTEN      0          383775     15787/dovecot       
tcp6       0      0 :::9900                 :::*                    LISTEN      0          383735     15787/dovecot       
```
I'm not sure anymore if that were all I modified getting it with that ipv6nat docker running. Though maybe it helps.
